### PR TITLE
feat: Sql runner job streaming

### PR DESCRIPTION
**Pull request by rephus**
Closes:  https://github.com/lightdash/lightdash/issues/10515

#### run (schedule job) 
Screenshot from 2024-07-01 14-52-21

#### Get job status 
Screenshot from 2024-07-01 15-16-36
#### On job error 
Screenshot from 2024-07-01 14-53-18

#### Results
Screenshot from 2024-07-01 15-37-36

### New e2e test
[Screencast from 01-07-24 16:50:04.webm](https://github.com/lightdash/lightdash/assets/1983672/375d5561-c143-4122-9c87-7681e8d75c75)

on max retries (reduced interval in snowflake) 

[Screencast from 01-07-24 16:33:40.webm](https://github.com/lightdash/lightdash/assets/1983672/40abf22b-fbbc-493a-8db8-d97e59128aec)

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

File: packages/backend/src/controllers/sqlRunnerController.ts [hunk 1]
```diff
@@ -69,4 +73,72 @@ export class SqlRunnerController extends BaseController {
...
+        @Request() req: express.Request,
+    ): Promise<any> {
+        this.setStatus(200);
+        this.setHeader('Content-Type', 'application/json');
+
+        const resultsFilePath = await this.services
+            .getProjectService()
+            .getResultsFile(req.user!, projectUuid, fileId);
+
+        // TODO set response headers
```
Comment (rephus) [inline packages/backend/src/controllers/sqlRunnerController.ts:121]: I've tried to add the right response headers, but they don't seem to work, I believe is a conflict with tsoa or sometihng. I can remove these comments and leave it for later.

File: packages/backend/src/controllers/sqlRunnerController.ts [hunk 2]
```diff
@@ -69,4 +73,72 @@ export class SqlRunnerController extends BaseController {
...
+      this.setStatus(200);
+      this.setHeader('Content-Type', mime.lookup(filePath));
+      this.setHeader('Content-Length', stat.size.toString());
+      // Removing this line will cause to not launch the download, just serve the file as it
+      this.setHeader('Content-Disposition', `attachment; filename=${fileName}`);
+      */
+        const mystream = fs.createReadStream(resultsFilePath);
+        const { res } = req;
+        if (res) {
+            mystream.pipe(res);
```
[the remark this answers is not in the document]
Comment (rephus) [reply packages/backend/src/controllers/sqlRunnerController.ts:135]: we could also implement this endpoint outside of tsoa

File: packages/backend/src/services/ProjectService/ProjectService.ts [hunk 3]
```diff
@@ -1,3 +1,4 @@
+import { NotFound } from '@aws-sdk/client-s3';
```
Comment (rephus) [inline packages/backend/src/services/ProjectService/ProjectService.ts:1]: I added the new streaming and save file functions into projectService, we should probably create a new `sqlRunnerService` or `streamingService` . If so, I'd like to do it on aseparate PR and move all relevant methods there too (eg: tables/fields)

File: packages/e2e/cypress/e2e/api/sqlRunner.cy.ts [hunk 4]
```diff
@@ -92,6 +93,97 @@ Object.entries(warehouseConnections).forEach(
...
+                        sql,
+                    }),
+                }).then((runResp) => {
+                    expect(runResp.status).to.eq(200);
+                    const { jobId } = runResp.body.results;
+
+                    const maxRetries = 20; // Snowflake is a bit slow compared to the others (normally takes ~3 seconds)
+
+                    // Poll request until job is `completed`
+                    const poll = (retries = 0) => {
```
Comment (rephus) [inline packages/e2e/cypress/e2e/api/sqlRunner.cy.ts:122]: this is a method to wrap the `job` status request, and call it recursively based on the `status` . 

Postgres (locally) takes <100ms to complete jobs, while snowflake takes ~3sec. I don't want to force cy.wait(5s) just to make it pass.

File: packages/backend/src/controllers/sqlRunnerController.ts (hunk 1 above)
Comment (rephus) [reply packages/backend/src/controllers/sqlRunnerController.ts]: https://github.com/lukeautry/tsoa/issues/800

File: packages/backend/src/services/ProjectService/ProjectService.ts [hunk 5]
```diff
@@ -1885,6 +1896,114 @@ export class ProjectService extends BaseService {
         return results;
     }
 
+    async streamResultsToLocalFile(
```
Comment (rephus) [inline packages/backend/src/services/ProjectService/ProjectService.ts:1899]: function that creates stores the download file in the database, creates a file stream and pass a function to `write`. 

We should create something similar for s3

File: packages/backend/src/controllers/sqlRunnerController.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (rephus) [reply packages/backend/src/controllers/sqlRunnerController.ts:124]: if we are going to keep this, I should probably return the readStream in `getResultsFile` instead of the path (I though tsoa should be able to read from a path like express.readFile)

File: packages/backend/src/services/ProjectService/ProjectService.ts [hunk 6]
```diff
@@ -1885,6 +1896,114 @@ export class ProjectService extends BaseService {
...
+    async streamSqlQueryIntoFile(
+        userUuid: string,
+        projectUuid: string,
+        sql: string,
+    ): Promise<string> {
+        const { organizationUuid } = await this.projectModel.getSummary(
+            projectUuid,
+        );
+
+        await this.analytics.track({
```
Comment (rephus) [inline packages/backend/src/services/ProjectService/ProjectService.ts:1945]: I'm keeping the same analytics track, should we make a new one ? / add context ?

File: packages/backend/src/services/ProjectService/ProjectService.ts [hunk 7]
```diff
@@ -1885,6 +1896,114 @@ export class ProjectService extends BaseService {
...
+            organization_uuid: organizationUuid,
+            user_uuid: userUuid,
+        };
+
+        // TODO upload to s3 if enabled
+        const fileId = await this.streamResultsToLocalFile(async (writter) => {
+            await warehouseClient.streamQuery(
+                sql,
+                async ({ rows, fields }) => {
+                    const formattedRows = formatRows(rows, {}); // TODO fields to itemmap
```
Comment (rephus) [inline packages/backend/src/services/ProjectService/ProjectService.ts:1967]: we should convert fields to itemmap to properly do formatting... I'm sure how to do it here (and efficiently)

File: packages/backend/src/services/ProjectService/ProjectService.ts [hunk 8]
```diff
@@ -1885,6 +1896,114 @@ export class ProjectService extends BaseService {
...
+                    formattedRows.forEach(writter);
```
Comment (rephus) [inline packages/backend/src/services/ProjectService/ProjectService.ts:1968]: write every line using the file streaming callback

File: packages/backend/src/services/ProjectService/ProjectService.ts [hunk 9]
```diff
@@ -1885,6 +1896,114 @@ export class ProjectService extends BaseService {
...
+                },
+                {
+                    tags: queryTags,
+                },
+            );
+        });
+
+        await sshTunnel.disconnect();
+        return `/api/v1/projects/${projectUuid}/sqlRunner/results/${fileId}`;
```
Comment (rephus) [inline packages/backend/src/services/ProjectService/ProjectService.ts:1977]: points to the new endpoint in sqlRunnerController to return the file.

File: packages/e2e/cypress/e2e/api/sqlRunner.cy.ts [hunk 10]
```diff
@@ -92,6 +93,97 @@ Object.entries(warehouseConnections).forEach(
...
+                                const { fileUrl } = resp.body.results.details;
+                                cy.request({
+                                    url: fileUrl,
+                                    method: 'GET',
+                                }).then((fileResp) => {
+                                    expect(fileResp.status).to.eq(200);
+                                    const lines = fileResp.body
+                                        .trim()
+                                        .split('\n');
+                                    const results = lines.map((line) =>
```
Comment (rephus) [inline packages/e2e/cypress/e2e/api/sqlRunner.cy.ts:145]: this is how we can consume a JSONL, not sure if there is a better way (JSON.load?)

File: packages/backend/src/services/ProjectService/ProjectService.ts [hunk 11]
```diff
@@ -1885,6 +1896,114 @@ export class ProjectService extends BaseService {
...
+        } finally {
+            writeStream.end(() => {
+                this.logger.debug('File has been saved.');
+            });
+        }
+
+        return downloadFileId;
+    }
+
+    async streamSqlQueryIntoFile(
```
Comment (rephus) [inline packages/backend/src/services/ProjectService/ProjectService.ts:1936]: perhaps we need to rename this function and add some code for s3

**Comment by rephus** [comment]
I still need to test the memory usage on large files :eyes:

File: packages/backend/src/services/ProjectService/ProjectService.ts (hunk 6 above)
Comment (ZeRego) [reply packages/backend/src/services/ProjectService/ProjectService.ts:1947]: We can add a new `usingStreaming: true` property in case we want to track usage during Beta.
btw, it's not a promise so we don't need the await.

File: packages/backend/src/services/ProjectService/ProjectService.ts (hunk 9 above)
Comment (ZeRego) [reply packages/backend/src/services/ProjectService/ProjectService.ts:1979]: We should return the full URL with domain as we will do for S3 files.
We also need to store the `itemMap` in the job state as we need that info in the FE.

File: packages/backend/src/services/ProjectService/ProjectService.ts (hunk 5 above)
Comment (ZeRego) [reply packages/backend/src/services/ProjectService/ProjectService.ts:1901]: in the ~database~ temp folder

**Comment by rephus** [comment]
Memory load testing (100k random results) 

### Saving files 
With streaming (not blocking API)
image
image

```
const fileId = await this.streamResultsToLocalFile(async (writter) => {
            await warehouseClient.streamQuery(
                sql,
                async ({ rows, fields }) => {
                    const formattedRows = formatRows(rows, {}); // TODO fields to itemmap
                    formattedRows.forEach(writter);
                },
                {
                    tags: queryTags,
                },
            );
        });
        ```

Without streaming (storing and formatting all rows in memory ) (blocking API)

Screenshot from 2024-07-03 10-01-35

```
const fileId = await this.streamResultsToLocalFile(async (writter) => {
            const { rows } = await await warehouseClient.runQuery(sql);
            const formattedRows = formatRows(rows, {}); // TODO fields to itemmap
            formattedRows.forEach(writter);

        });
        ```

**Comment by rephus** [comment]
Memory load testing (50k rows) 

## Streaming results 

### With read buffer pipe 
Screenshot from 2024-07-03 10-20-02
Screenshot from 2024-07-03 10-20-11

```

 if (res) {
            readStream.pipe(res);
            await new Promise<void>((resolve, reject) => {
                readStream.on('end', () => {
                    res.end();
                    console.info('after readStream', process.memoryUsage())
                    resolve();
                });
            });
        }
```

File: packages/e2e/cypress/e2e/api/sqlRunner.cy.ts [hunk 12]
```diff
@@ -83,6 +88,178 @@ Object.entries(warehouseConnections).forEach(
...
+        // This SQL generates 50k random results on Postgres
+        const sql = `
+            SELECT
+                gs.id AS payment_id,
+                (1000 + random() * 10000)::int AS amount,  -- Random amount between 1000 and 11000
+                (current_date - (random() * 365)::int)::date AS payment_date,  -- Random date within the last year
+                md5(random()::text) AS payment_method,  -- Random payment reference
+                (random() * 100000)::int AS customer_id  -- Random customer_id between 0 and 100000
+            FROM
+                generate_series(1, 50000) AS gs(id)  -- Generate 50k rows
```
Comment (ZeRego) [inline packages/e2e/cypress/e2e/api/sqlRunner.cy.ts:205]: Note that I'm about to enforce a limit of 5k on the SQL runner results. 
Enforce a limit in the SQL runner [#10238](https://github.com/lightdash/lightdash/issues/10238)

**Review by ZeRego — APPROVED** [review]
